src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,144 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (head == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        Node<T> currentNode = node(index);
+        Node<T> prevNode = currentNode.prev;
+        Node<T> newNode = new Node<>(prevNode, value, currentNode);
+        currentNode.prev = newNode;
+        if (prevNode == null) {
+            head = newNode;
+        } else {
+            prevNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = node(index);
+        T firstValue = node.value;
+        node.value = value;
+        return firstValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> removeObject = head;
+        for (int i = 0; i < size; i++) {
+            if (object != null && object.equals(removeObject.value)
+                    || removeObject.value == object) {
+                remove(i);
+                return true;
+            }
+            removeObject = removeObject.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index: " + index + " for size: " + size);
+        }
+    }
+
+    private T unlink(Node<T> node) {
+        final T deletedNode = node.value;
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+        size--;
+        return deletedNode;
+    }
+
+    private Node<T> node(int index) {
```

Comment:
```suggestion
    private Node<T> getNode(int index) {
```

https://mate-academy.github.io/style-guides/java/java.html#s5.2.3-method-names